azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,689 @@
+            return
+
+        try:
+            logger.info("Move process {0} into cgroups for extension {1}".format(pid, name))
+            CGroups.for_extension(name).add(pid)
+        except Exception as ex:
+            logger.warn("Unable to move process {0} into cgroups for extension {1}: {2}".format(pid, name, ex))
+
+    @staticmethod
+    def get_my_cgroup_path(hierarchy_id):
+        """
+        Get the cgroup path "suffix" for this process for the given hierarchy ID. The leading "/" is always stripped,
+        so the suffix is suitable for passing to os.path.join(). (If the process is in the root cgroup, an empty
+        string is returned, and os.path.join() will still do the right thing.)
+
+        :param hierarchy_id: str
+        :return: str
+        """
+        cgroup_paths = fileutil.read_file("/proc/self/cgroup")
+        for entry in cgroup_paths.splitlines():
+            fields = entry.split(':')
+            if fields[0] == hierarchy_id:
+                return fields[2].lstrip(os.path.sep)
+        raise CGroupsException("This process belongs to no cgroup for hierarchy ID {0}".format(hierarchy_id))
+
+    @staticmethod
+    def get_hierarchy_id(hierarchy):
+        """
+        Get the cgroups hierarchy ID for a given hierarchy name
+
+        :param hierarchy:
+        :return: str
+        """
+        cgroup_states = fileutil.read_file("/proc/cgroups")
+        for entry in cgroup_states.splitlines():
+            fields = entry.split('\t')
+            if fields[0] == hierarchy:
+                return fields[1]
+        raise CGroupsException("Cgroup hierarchy {0} not found in /proc/cgroups".format(hierarchy))
+
+    @staticmethod
+    def get_my_cgroup_folder(hierarchy):
+        """
+        Find the path of the cgroup in which this process currently lives for the given hierarchy.
+
+        :param hierarchy: str
+        :return: str
+        """
+        hierarchy_id = CGroups.get_hierarchy_id(hierarchy)
+        return os.path.join(BASE_CGROUPS, hierarchy, CGroups.get_my_cgroup_path(hierarchy_id))
+
+    def _get_cgroup_file(self, hierarchy, file_name):
+        return os.path.join(self.cgroups[hierarchy], file_name)
+
+    @staticmethod
+    def _convert_cpu_limit_to_fraction(value):
+        """
+        Convert a CPU limit from percent (e.g. 50 meaning 50%) to a decimal fraction (0.50).
+        :return: Fraction of one CPU to be made available (e.g. 0.5 means half a core)
+        :rtype: float
+        """
+        try:
+            limit = float(value)
+        except ValueError:
+            raise CGroupsException('CPU Limit must be convertible to a float')
+
+        if limit <= float(0) or limit > float(CGroups.get_num_cores() * 100):
+            raise CGroupsException('CPU Limit must be between 0 and 100 * numCores')
+
+        return limit / 100.0
+
+    def get_file(self, hierarchy, file_name):
+        """
+        Retrieve the value of a parameter from a hierarchy.
```

Comment:
The method is called get_file, but that doesn't match the comment or code.  Is read_file or get_file_contents better? Is this method only intended to be call by get_parameter?

Reply:
CGroups.get_file_contents() it is. Any holder of a CGroups object can legitimately call this; cgroups is so wildly varied that it's hard to know. I am pretty sure at least one DCR test for cgroups relies on this method being public.